plugins/soundsourcem4a/audiosourcem4a.cpp

```diff
@@ -0,0 +1,307 @@
+#include "audiosourcem4a.h"
+
+#include "sampleutil.h"
+
+#ifdef __WINDOWS__
+#include <io.h>
+#include <fcntl.h>
+#endif
+
+#ifdef _MSC_VER
+#define S_ISDIR(mode) (mode & _S_IFDIR)
+#define strcasecmp stricmp
+#define strncasecmp strnicmp
+#endif
+
+#ifdef __M4AHACK__
+typedef uint32_t SAMPLERATE_TYPE;
+#else
+typedef unsigned long SAMPLERATE_TYPE;
+#endif
+
+namespace Mixxx {
+
+namespace
+{
+
+// AAC: "... each block decodes to 1024 time-domain samples."
+const AudioSource::size_type kFramesPerSampleBlock = 1024;
+
+// MP4SampleId is 1-based
+const MP4SampleId kMinSampleId = 1;
+
+// Decoding will be restarted one or more blocks of samples
+// before the actual position to avoid audible glitches.
+// Two blocks of samples seems to be enough here.
+const MP4SampleId kSampleIdPrefetchCount = 2;
+
+MP4TrackId findAacTrackId(MP4FileHandle hFile) {
+    const MP4TrackId maxTrackId = MP4GetNumberOfTracks(hFile, NULL, 0);
+    for (MP4TrackId trackId = 1; trackId <= maxTrackId; ++trackId) {
+        const char* trackType = MP4GetTrackType(hFile, trackId);
+        if ((NULL == trackType) || !MP4_IS_AUDIO_TRACK_TYPE(trackType)) {
+            continue;
+        }
+        const char* mediaDataName = MP4GetTrackMediaDataName(hFile, trackId);
+        if (0 != strcasecmp(mediaDataName, "mp4a")) {
+            continue;
+        }
+        const u_int8_t audioType = MP4GetTrackEsdsObjectTypeId(hFile, trackId);
+        if (MP4_INVALID_AUDIO_TYPE == audioType) {
+            continue;
+        }
+        if (MP4_MPEG4_AUDIO_TYPE == audioType) {
+            const u_int8_t audioMpeg4Type = MP4GetTrackAudioMpeg4Type(hFile,
+                    trackId);
+            if (MP4_IS_MPEG4_AAC_AUDIO_TYPE(audioMpeg4Type)) {
+                return trackId;
+            }
+        } else if (MP4_IS_AAC_AUDIO_TYPE(audioType)) {
+            return trackId;
+        }
+    }
+    return MP4_INVALID_TRACK_ID;
+}
+
+}
+
+AudioSourceM4A::AudioSourceM4A()
+        : m_hFile(MP4_INVALID_FILE_HANDLE), m_trackId(
```

Comment:
Code style nit pick: It applies to the coding rules, but please use one line for every initializer, I am confused.